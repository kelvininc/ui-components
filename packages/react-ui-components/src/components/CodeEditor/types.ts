import { Monaco } from '@monaco-editor/react';
import type { editor } from 'monaco-editor';
import { ForwardedRef, FunctionComponent } from 'react';

export type CodeEditor = editor.IStandaloneCodeEditor;
export type CodeInstance = Monaco;
export type OnCodeEditorChange = (value: string | undefined) => void;
export type CodeEditorOptions = editor.IGlobalEditorOptions & editor.IEditorOptions;
export type CodeDiffEditorOptions = editor.IGlobalEditorOptions & editor.IDiffEditorOptions;

export type CodeEditorTheme = editor.IStandaloneThemeData;

export enum ECodeEditorTheme {
	Dark = 'vs-dark',
	Light = 'light',
	Custom = 'custom'
}

export interface ICodeEditorProps {
	/** Use this property to pass a starting value to the editor. */
	code?: string;
	/** Use this property to define the language mode of the editor. */
	language?: string;
	/** Use this property to define the theme of the editor. */
	theme?: ECodeEditorTheme;
	/** Use this property to define and set a custom theme on the editor */
	customTheme?: CodeEditorTheme;
	/** Use this property to define options on the editor */
	customOptions?: CodeEditorOptions;
	/** Use this property to define a placeholder component to display when the editor is loading. */
	LoadingComponent?: FunctionComponent;
	/** Use this property to pass a ref to the editor. */
	forwardedRef?: ForwardedRef<CodeEditor>;
	/** Use this property to pass a ref to the editor instance. */
	instanceRef?: ForwardedRef<CodeInstance>;
	/** Use this property to pass a callback function for when the value (`code`) changes. */
	onChange?: OnCodeEditorChange;
}

export interface ICodeDiffEditorProps extends Omit<ICodeEditorProps, 'code' | 'customOptions'> {
	/** Use this property to pass the original code value to the editor. */
	originalCode?: string;
	/** Use this property to pass the modified code value to the editor. */
	modifiedCode?: string;
	/** Use this property to define options on the editor */
	customOptions?: CodeDiffEditorOptions;
}
