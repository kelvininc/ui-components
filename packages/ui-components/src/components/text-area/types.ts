import { EventEmitter } from '@stencil/core';
import { EIconName } from '../icon/icon.types';

export interface ITextArea {
	/** (optional) Icon to show to the left of the text field */
	icon?: EIconName;
	/** (optional) The text to show inside the text area */
	text?: string;
	/** (optional) The placeholder to show in the text area */
	placeholder?: string;
	/** (optional) The maximum number of characters allowed */
	maxCharLength?: number;
	/** (optional) If `true` the chars counter is displayed. Default: `true` */
	counter?: boolean;
	/** (optional) If `true` the text area is disabled. Default: `false`. */
	disabled?: boolean;
}

export interface ITextAreaEvents {
	/** Emits the current text when there's a change */
	textChange: EventEmitter<string>;
}
