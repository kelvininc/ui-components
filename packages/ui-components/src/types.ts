export * from './utils/types';

// Component type exports
export * from './components/absolute-time-picker/absolute-time-picker.types';
export * from './components/absolute-time-picker-dropdown/absolute-time-picker-dropdown.types';
export * from './components/absolute-time-picker-dropdown-input/absolute-time-picker-dropdown-input.types';
export * from './components/action-button/action-button.types';
export * from './components/action-button-icon/action-button.types';
export * from './components/action-button-split/action-button-split.types';
export * from './components/action-button-text/action-button-text.types';
export * from './components/alert/alert.types';
export * from './components/badge/badge.types';
export * from './components/breadcrumb-item/breadcrumb-item.types';
export * from './components/calendar/calendar.types';
export * from './components/calendar-day/calendar-day.types';
export * from './components/checkbox/checkbox.types';
export * from './components/copy-to-clipboard/copy-to-clipboard.types';
export * from './components/date-time-input/date-time-input.types';
export * from './components/description-list/description-list.types';
export * from './components/dropdown/dropdown.types';
export * from './components/dropdown-base/dropdown-base.types';
export * from './components/icon/icon.types';
export * from './components/illustration/illustration.types';
export * from './components/illustration-message/illustration-message.types';
export * from './components/info-label/info-label.types';
export * from './components/input-wrapper/input-wrapper.types';
export * from './components/link/link.types';
export * from './components/modal/modal.types';
export * from './components/multi-select-dropdown/multi-select-dropdown.types';
export * from './components/portal/portal.types';
export * from './components/radio-list/radio-list.types';
export * from './components/radio-list-item/radio-list-item.types';
export * from './components/range/range.types';
export * from './components/relative-time-picker/relative-time-picker.types';
export * from './components/search/search.types';
export * from './components/select/select.types';
export * from './components/select-create-option/select-create-option.types';
export * from './components/select-multi-options/select-multi-options.types';
export * from './components/select-option/select-option.types';
export * from './components/single-select-dropdown/single-select-dropdown.types';
export * from './components/step-bar/step-bar.types';
export * from './components/step-indicator/step-indicator.types';
export * from './components/step-progress-bar/step-progress-bar.types';
export * from './components/summary-card/summary-card.types';
export * from './components/switch-button/switch-button.types';
export * from './components/tab-navigation/tab-navigation.types';
export * from './components/tag-alarm/tag-alarm.types';
export * from './components/tag-status/tag-status.types';
export * from './components/text-field/text-field.types';
export * from './components/time-picker/time-picker.types';
export * from './components/time-picker-select-option/time-picker-select-option.types';
export * from './components/toaster/toaster.types';
export * from './components/toggle-button/toggle-button.types';
export * from './components/toggle-button-group/toggle-button-group.types';
export * from './components/toggle-switch/toggle-switch.types';
export * from './components/toggle-tip/toggle-tip.types';
export * from './components/tooltip/tooltip.types';
export * from './components/tooltip-text/tooltip-text.types';
export * from './components/tree/tree.types';
export * from './components/tree-item/tree-item.types';
export * from './components/virtualized-list/virtualized-list.types';
export * from './components/wizard/wizard.types';
export * from './components/wizard-footer/wizard-footer.types';
export * from './components/wizard-header/wizard-header.types';
