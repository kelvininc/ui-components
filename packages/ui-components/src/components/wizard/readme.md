# _<kv-wizard>_



<!-- Auto Generated Below -->


## Usage

### React

```tsx
import React from 'react';

import { IWizardStep, KvWizard } from '@kelvininc/react-ui-components';

export const KvWizardExample: React.FC = () => {
	const steps: IWizardStep[]  = [
		{
			title: 'Info' ,
			allowGoBack: true
		},
		{
			enabled: 'Configuration',
			allowGoBack: true
		}
	];

	const onGoToStep = ({ detail }: CustomEvent<number>) => {};
	const onSubmitClick = (ev: CustomEvent<void>) => {};
	const onCancelClick = (ev: CustomEvent<void>) => {};

	return (
		<KvWizardFooter
			steps={steps}
			currentStep={0}
			hasError={false}
			showStepBar={true}
			showHeader={true}
			submitBtnLabel="Deploy"
			onGoToStep={onGoToStep}
			onSubmitClick={onSubmitClick}
			onCancelClick={onCancelClick} />
	);
};
```



## Properties

| Property                   | Attribute            | Description                                                                         | Type                                                                                            | Default     |
| -------------------------- | -------------------- | ----------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------- | ----------- |
| `completeBtnLabel`         | `complete-btn-label` | (optional) A label to show on the last step button. Default: 'Submit'               | `string`                                                                                        | `undefined` |
| `currentStep` _(required)_ | `current-step`       | (required) Defines the current step index                                           | `number`                                                                                        | `undefined` |
| `currentStepState`         | `current-step-state` | (optional) Defines the current step state. Only the success state allows to proceed | `{ state: EStepState.Error; error?: string; } \| { state: EStepState.Success; error?: never; }` | `undefined` |
| `disabled`                 | `disabled`           | (optional) Defines if the wizard navigation is disabled. Default: false             | `boolean`                                                                                       | `false`     |
| `showHeader`               | `show-header`        | (optional) Defines if the header should render. Default: true                       | `boolean`                                                                                       | `true`      |
| `showStepBar`              | `show-step-bar`      | (optional) Defines if the step bar should render. Default: true                     | `boolean`                                                                                       | `true`      |
| `steps` _(required)_       | `steps`              | (required) Defines the wizard steps                                                 | `IWizardStep[]`                                                                                 | `undefined` |


## Events

| Event           | Description                                         | Type                      |
| --------------- | --------------------------------------------------- | ------------------------- |
| `cancelClick`   | Fires when a cancel button is clicked               | `CustomEvent<MouseEvent>` |
| `completeClick` | Fires when a complete button is clicked             | `CustomEvent<MouseEvent>` |
| `goToStep`      | Fires when it's necessary to go to a different step | `CustomEvent<number>`     |


## CSS Custom Properties

| Name                     | Description                                       |
| ------------------------ | ------------------------------------------------- |
| `--wizard-footer-height` | defines how much space the footer should take up  |
| `--wizard-header-height` | defines how much space the header should take up  |
| `--wizard-stepper-width` | defines how much space the stepper should take up |


## Dependencies

### Depends on

- [kv-wizard-header](../wizard-header)
- [kv-wizard-footer](../wizard-footer)

### Graph
```mermaid
graph TD;
  kv-wizard --> kv-wizard-header
  kv-wizard --> kv-wizard-footer
  kv-wizard-header --> kv-toggle-tip
  kv-wizard-header --> kv-icon
  kv-toggle-tip --> kv-portal
  kv-toggle-tip --> kv-tooltip-text
  kv-wizard-footer --> kv-step-bar
  kv-wizard-footer --> kv-action-button-text
  kv-wizard-footer --> kv-tooltip
  kv-step-bar --> kv-step-progress-bar
  kv-step-bar --> kv-step-indicator
  kv-step-indicator --> kv-icon
  kv-action-button-text --> kv-action-button
  kv-action-button-text --> kv-icon
  kv-tooltip --> kv-portal
  kv-tooltip --> kv-tooltip-text
  style kv-wizard fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


