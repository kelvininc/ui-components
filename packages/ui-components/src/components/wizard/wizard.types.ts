import { EventEmitter } from '@stencil/core';

export interface IWizardStep {
	title: string;
	cancelable?: boolean;
	allowGoBack?: boolean;
	tip?: string;
}

export interface IWizard {
	/** (required) Defines the wizard steps */
	steps: IWizardStep[];
	/** (required) Defines the current step index */
	currentStep: number;
	/** (optional) Defines the current step state. Only the success state allows to proceed */
	currentStepState?: StepState;
	/** (optional) Defines if the header should render. Default: true */
	showHeader?: boolean;
	/** (optional) Defines if the step bar should render. Default: true */
	showStepBar?: boolean;
	/** (optional) Defines if the wizard navigation is disabled. Default: false */
	disabled?: boolean;
	/** (optional) A label to show on the last step button. Default: 'Submit' */
	completeBtnLabel?: string;
}

export interface IWizardEvents {
	/** Fires when it's necessary to go to a different step */
	goToStep: EventEmitter<number>;
	/** Fires when a complete button is clicked */
	completeClick: EventEmitter<MouseEvent>;
	/** Fires when a cancel button is clicked */
	cancelClick: EventEmitter<MouseEvent>;
}

export type StepState =
	| {
			state: EStepState.Success;
			error?: never;
	  }
	| {
			state: EStepState.Error;
			error?: string;
	  };

export enum EStepState {
	Success = 'success',
	Error = 'error'
}
