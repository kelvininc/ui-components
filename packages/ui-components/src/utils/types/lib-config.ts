export interface UIComponentsConfig {
	baseAssetsUrl?: string;
	symbolsFileName?: string;
}

declare global {
	interface Window {
		KvUiComponents?: {
			config?: UIComponentsConfig;
		};
	}
}
