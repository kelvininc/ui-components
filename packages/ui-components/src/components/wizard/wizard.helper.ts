import { isEmpty, isNil } from 'lodash-es';
import { IStepBarStep } from '../step-bar/step-bar.types';
import { IWizardFooter } from '../wizard-footer/wizard-footer.types';
import { IWizardHeader } from '../wizard-header/wizard-header.types';
import { EStepState, IWizardStep, StepState } from './wizard.types';

export const buildHeaderConfig = (steps?: IWizardStep[], currentStep?: number): IWizardHeader => {
	if (isEmpty(steps) || isNil(currentStep) || currentStep < 0 || currentStep > steps.length - 1) {
		return null;
	}

	return {
		label: `Step ${currentStep + 1}`,
		description: steps[currentStep].title,
		tip: steps[currentStep].tip
	};
};

export const buildFooterConfig = (steps?: IWizardStep[], currentStep?: number, currentStepState?: StepState, disabled = false): IWizardFooter => {
	if (isEmpty(steps) || isNil(currentStep) || currentStep < 0 || currentStep > steps.length - 1) {
		return null;
	}

	const stepsConfig: IStepBarStep[] = steps.map((step, index) => {
		return {
			stepKey: step.title,
			enabled: !disabled && ((index === currentStep + 1 && currentStepState?.state === EStepState.Success) || (index < currentStep && steps[index + 1].allowGoBack)),
			active: index <= currentStep,
			hasError: index === currentStep && currentStepState?.state === EStepState.Error
		};
	});

	return {
		steps: stepsConfig,
		currentStep,
		hasError: false,
		showPrevBtn: steps[currentStep].allowGoBack ?? false,
		prevEnabled: !disabled && currentStep > 0,
		showNextBtn: currentStep < steps.length - 1,
		nextEnabled: !disabled && currentStepState?.state === EStepState.Success,
		nextTooltip: currentStepState?.error,
		showCancelBtn: steps[currentStep].cancelable ?? false,
		cancelEnabled: true,
		showCompleteBtn: currentStep === steps.length - 1,
		completeEnabled: !disabled && currentStepState?.state === EStepState.Success,
		completeTooltip: currentStepState?.error,
		progressPercentage: currentStep * (100 / (steps.length - 1))
	};
};
