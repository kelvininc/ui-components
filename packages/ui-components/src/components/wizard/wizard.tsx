import { Component, Event, EventEmitter, Prop, h, State, Watch, Listen } from '@stencil/core';
import { IWizardFooter } from '../wizard-footer/wizard-footer.types';
import { IWizardHeader } from '../wizard-header/wizard-header.types';
import { buildFooterConfig, buildHeaderConfig } from './wizard.helper';
import { IWizard, IWizardEvents, IWizardStep, StepState, EStepState } from './wizard.types';

@Component({
	tag: 'kv-wizard',
	styleUrl: 'wizard.scss',
	shadow: true
})
export class KvWizard implements IWizard, IWizardEvents {
	/** @inheritdoc */
	@Prop({ reflect: true }) steps!: IWizardStep[];
	/** @inheritdoc */
	@Prop({ reflect: true }) currentStep!: number;
	/** @inheritdoc */
	@Prop({ reflect: true }) currentStepState?: StepState;
	/** @inheritdoc */
	@Prop({ reflect: true }) showHeader: boolean = true;
	/** @inheritdoc */
	@Prop({ reflect: true }) showStepBar: boolean = true;
	/** @inheritdoc */
	@Prop({ reflect: true }) disabled: boolean = false;
	/** @inheritdoc */
	@Prop({ reflect: true }) completeBtnLabel?: string;

	/** @inheritdoc */
	@Event() goToStep: EventEmitter<number>;
	/** @inheritdoc */
	@Event() completeClick: EventEmitter<MouseEvent>;
	/** @inheritdoc */
	@Event() cancelClick: EventEmitter<MouseEvent>;

	/** Internal header state */
	@State() currentHeader: IWizardHeader;
	/** Internal footer state */
	@State() currentFooter: IWizardFooter;

	/** Watch the `steps` property and update internal state accordingly */
	@Watch('steps')
	stepsChangeHandler(newValue: IWizardStep[]) {
		this.currentHeader = buildHeaderConfig(newValue, this.currentStep);
		this.currentFooter = buildFooterConfig(newValue, this.currentStep, this.currentStepState, this.disabled);
	}
	/** Watch the `currentStep` property and update internal state accordingly */
	@Watch('currentStep')
	currentStepChangeHandler(newValue: number) {
		this.currentHeader = buildHeaderConfig(this.steps, newValue);
		this.currentFooter = buildFooterConfig(this.steps, newValue, this.currentStepState, this.disabled);
	}
	/** Watch the `currentStepState` property and update internal state accordingly */
	@Watch('currentStepState')
	hasErrorStepChangeHandler(newValue: StepState) {
		this.currentFooter = buildFooterConfig(this.steps, this.currentStep, newValue, this.disabled);
	}

	componentWillLoad() {
		this.currentHeader = buildHeaderConfig(this.steps, this.currentStep);
		this.currentFooter = buildFooterConfig(this.steps, this.currentStep, this.currentStepState, this.disabled);
	}

	onPrevClick = () => {
		this.goToStep.emit(this.currentStep - 1);
	};

	onNextClick = () => {
		this.goToStep.emit(this.currentStep + 1);
	};

	onStepClick = ({ detail }: CustomEvent<number>) => {
		this.goToStep.emit(detail);
	};

	@Listen('keydown', { target: 'document' })
	handleKeyDown(event: KeyboardEvent) {
		if (event.key === 'Enter') {
			// Don't handle if focus is on a textarea
			const activeElement = document.activeElement;
			if (activeElement instanceof HTMLTextAreaElement || event.target instanceof HTMLTextAreaElement) {
				return;
			}

			// Don't handle if wizard is disabled
			if (this.disabled) {
				return;
			}

			// Don't handle if current step has errors
			if (this.currentStepState?.state === EStepState.Error) {
				return;
			}

			// Don't handle if the component hasn't fully initialized
			if (!this.currentFooter) {
				return;
			}

			event.preventDefault();

			const { completeEnabled, showCompleteBtn, nextEnabled } = this.currentFooter;

			if (showCompleteBtn) {
				// Don't handle if complete button is disabled
				if (!completeEnabled) {
					return;
				}

				// Create a synthetic mouse event for consistency with existing API
				const syntheticEvent = new MouseEvent('click', {
					bubbles: true,
					cancelable: true
				});

				return this.completeClick.emit(syntheticEvent);
			}

			if (!nextEnabled) {
				return;
			}

			// Move to next step
			return this.goToStep.emit(this.currentStep + 1);
		}
	}

	render() {
		return (
			<div class="wizard">
				{this.showHeader && (
					<div class="wizard-header">
						{this.currentHeader && (
							<kv-wizard-header {...this.currentHeader}>
								<slot slot="additional-header-actions" name="additional-header-actions" />
							</kv-wizard-header>
						)}
					</div>
				)}
				<div class={{ 'wizard-content': true, 'has-header': this.showHeader }}>
					<slot name="step-content"></slot>
				</div>
				<div class="wizard-footer">
					<kv-wizard-footer
						onPrevClick={this.onPrevClick}
						onNextClick={this.onNextClick}
						onStepClick={this.onStepClick}
						showStepBar={this.showStepBar}
						completeBtnLabel={this.completeBtnLabel}
						exportparts="footer-actions-container"
						{...this.currentFooter}
					>
						<slot slot="additional-actions" name="additional-actions" />
					</kv-wizard-footer>
				</div>
			</div>
		);
	}
}
