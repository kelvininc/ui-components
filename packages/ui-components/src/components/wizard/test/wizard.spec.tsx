import { SpecPage, newSpecPage } from '@stencil/core/testing';
import { h } from '@stencil/core';
import { MOCK_STEPS } from './wizard.mock';
import { KvWizard } from '../wizard';
import { EStepState } from '../wizard.types';

describe('Wizard (unit tests)', () => {
	let page: SpecPage;
	let comp: KvWizard;

	describe("when current step don't have state", () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvWizard],
				template: () => <kv-wizard steps={MOCK_STEPS} currentStep={1} />
			});
			comp = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});

		it('should initialize the header props with the correct value', () => {
			expect(comp.currentHeader).toEqual({
				label: 'Step 2',
				description: 'Configuration'
			});
		});

		it('should initialize the footer props with the correct value', () => {
			expect(comp.currentFooter).toEqual({
				steps: [
					{
						stepKey: 'Info',
						enabled: true,
						active: true,
						hasError: false
					},
					{
						stepKey: 'Configuration',
						enabled: false,
						active: true,
						hasError: false
					},
					{
						stepKey: 'Confirmation',
						enabled: false,
						active: false,
						hasError: false
					}
				],
				currentStep: 1,
				hasError: false,
				showPrevBtn: true,
				prevEnabled: true,
				showNextBtn: true,
				nextEnabled: false,
				showCancelBtn: false,
				cancelEnabled: true,
				showCompleteBtn: false,
				completeEnabled: false,
				progressPercentage: 50
			});
		});
	});

	describe('when current step have errors', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvWizard],
				template: () => <kv-wizard steps={MOCK_STEPS} currentStep={1} currentStepState={{ state: EStepState.Error, error: 'Unable to proceed to the next step' }} />
			});
			comp = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});

		it('should initialize the header props with the correct value', () => {
			expect(comp.currentHeader).toEqual({
				label: 'Step 2',
				description: 'Configuration'
			});
		});

		it('should initialize the footer props with the correct value', () => {
			expect(comp.currentFooter).toEqual({
				steps: [
					{
						stepKey: 'Info',
						enabled: true,
						active: true,
						hasError: false
					},
					{
						stepKey: 'Configuration',
						enabled: false,
						active: true,
						hasError: true
					},
					{
						stepKey: 'Confirmation',
						enabled: false,
						active: false,
						hasError: false
					}
				],
				currentStep: 1,
				hasError: false,
				showPrevBtn: true,
				prevEnabled: true,
				showNextBtn: true,
				nextEnabled: false,
				nextTooltip: 'Unable to proceed to the next step',
				showCancelBtn: false,
				cancelEnabled: true,
				showCompleteBtn: false,
				completeEnabled: false,
				completeTooltip: 'Unable to proceed to the next step',
				progressPercentage: 50
			});
		});
	});

	describe('when current step have success state', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvWizard],
				template: () => <kv-wizard steps={MOCK_STEPS} currentStep={1} currentStepState={{ state: EStepState.Success }} />
			});
			comp = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});

		it('should initialize the header props with the correct value', () => {
			expect(comp.currentHeader).toEqual({
				label: 'Step 2',
				description: 'Configuration'
			});
		});

		it('should initialize the footer props with the correct value', () => {
			expect(comp.currentFooter).toEqual({
				steps: [
					{
						stepKey: 'Info',
						enabled: true,
						active: true,
						hasError: false
					},
					{
						stepKey: 'Configuration',
						enabled: false,
						active: true,
						hasError: false
					},
					{
						stepKey: 'Confirmation',
						enabled: true,
						active: false,
						hasError: false
					}
				],
				currentStep: 1,
				hasError: false,
				showPrevBtn: true,
				prevEnabled: true,
				showNextBtn: true,
				nextEnabled: true,
				showCancelBtn: false,
				cancelEnabled: true,
				showCompleteBtn: false,
				completeEnabled: true,
				progressPercentage: 50
			});
		});

		describe('and step 2 is clicked on step bar', () => {
			beforeEach(() => {
				jest.spyOn(comp.goToStep, 'emit');
				comp.onStepClick({ detail: 2 } as CustomEvent<number>);
			});

			it('should emit to goToStep 2', () => {
				expect(comp.goToStep.emit).toHaveBeenCalledWith(2);
			});
		});

		describe('and previous button is clicked', () => {
			beforeEach(() => {
				jest.spyOn(comp.goToStep, 'emit');
				comp.onPrevClick();
			});

			it('should emit to goToStep 1', () => {
				expect(comp.goToStep.emit).toHaveBeenCalledWith(0);
			});
		});

		describe('and next button is clicked', () => {
			beforeEach(() => {
				jest.spyOn(comp.goToStep, 'emit');
				comp.onNextClick();
			});

			it('should emit to goToStep 1', () => {
				expect(comp.goToStep.emit).toHaveBeenCalledWith(2);
			});
		});
	});

	describe('when current step is the first one', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvWizard],
				template: () => <kv-wizard steps={MOCK_STEPS} currentStep={0} currentStepState={{ state: EStepState.Success }} />
			});
			comp = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});

		it('should initialize the header props with the correct value', () => {
			expect(comp.currentHeader).toEqual({
				label: 'Step 1',
				description: 'Info'
			});
		});

		it('should initialize the footer props with the correct value', () => {
			expect(comp.currentFooter).toEqual({
				steps: [
					{
						stepKey: 'Info',
						enabled: false,
						active: true,
						hasError: false
					},
					{
						stepKey: 'Configuration',
						enabled: true,
						active: false,
						hasError: false
					},
					{
						stepKey: 'Confirmation',
						enabled: false,
						active: false,
						hasError: false
					}
				],
				currentStep: 0,
				hasError: false,
				showPrevBtn: false,
				prevEnabled: false,
				showNextBtn: true,
				nextEnabled: true,
				showCancelBtn: true,
				cancelEnabled: true,
				showCompleteBtn: false,
				completeEnabled: true,
				progressPercentage: 0
			});
		});
	});

	describe('when current step is the last one', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvWizard],
				template: () => <kv-wizard steps={MOCK_STEPS} currentStep={2} currentStepState={{ state: EStepState.Success }} completeBtnLabel="Deploy" />
			});
			comp = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});

		it('should initialize the header props with the correct value', () => {
			expect(comp.currentHeader).toEqual({
				label: 'Step 3',
				description: 'Confirmation'
			});
		});

		it('should initialize the footer props with the correct value', () => {
			expect(comp.currentFooter).toEqual({
				steps: [
					{
						stepKey: 'Info',
						enabled: true,
						active: true,
						hasError: false
					},
					{
						stepKey: 'Configuration',
						enabled: true,
						active: true,
						hasError: false
					},
					{
						stepKey: 'Confirmation',
						enabled: false,
						active: true,
						hasError: false
					}
				],
				currentStep: 2,
				hasError: false,
				showPrevBtn: true,
				prevEnabled: true,
				showNextBtn: false,
				nextEnabled: true,
				showCancelBtn: false,
				cancelEnabled: true,
				showCompleteBtn: true,
				completeEnabled: true,
				progressPercentage: 100
			});
		});

		describe('and previous button is clicked', () => {
			beforeEach(() => {
				jest.spyOn(comp.goToStep, 'emit');
				comp.onPrevClick();
			});

			it('should emit to goToStep 1', () => {
				expect(comp.goToStep.emit).toHaveBeenCalledWith(1);
			});
		});
	});

	describe('when showHeader is false', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvWizard],
				template: () => <kv-wizard steps={MOCK_STEPS} currentStep={1} showHeader={false} />
			});
			comp = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});

	describe('when showStepBar is false', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvWizard],
				template: () => <kv-wizard steps={MOCK_STEPS} currentStep={1} showStepBar={false} />
			});
			comp = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});
});
