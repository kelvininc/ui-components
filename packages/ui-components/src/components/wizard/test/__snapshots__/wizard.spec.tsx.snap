// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wizard (unit tests) when current step don't have state should match the snapshot 1`] = `
<kv-wizard current-step="1" show-header="" show-step-bar="">
  <template shadowrootmode="open">
    <div class="wizard">
      <div class="wizard-header">
        <kv-wizard-header description="Configuration" label="Step 2">
          <slot name="additional-header-actions" slot="additional-header-actions"></slot>
        </kv-wizard-header>
      </div>
      <div class="has-header wizard-content">
        <slot name="step-content"></slot>
      </div>
      <div class="wizard-footer">
        <kv-wizard-footer cancelenabled="" currentstep="1" exportparts="footer-actions-container" prevenabled="" progresspercentage="50" shownextbtn="" showprevbtn="" showstepbar="">
          <slot name="additional-actions" slot="additional-actions"></slot>
        </kv-wizard-footer>
      </div>
    </div>
  </template>
</kv-wizard>
`;

exports[`Wizard (unit tests) when current step have errors should match the snapshot 1`] = `
<kv-wizard current-step="1" current-step-state="error" show-header="" show-step-bar="">
  <template shadowrootmode="open">
    <div class="wizard">
      <div class="wizard-header">
        <kv-wizard-header description="Configuration" label="Step 2">
          <slot name="additional-header-actions" slot="additional-header-actions"></slot>
        </kv-wizard-header>
      </div>
      <div class="has-header wizard-content">
        <slot name="step-content"></slot>
      </div>
      <div class="wizard-footer">
        <kv-wizard-footer cancelenabled="" currentstep="1" exportparts="footer-actions-container" prevenabled="" progresspercentage="50" shownextbtn="" showprevbtn="" showstepbar="">
          <slot name="additional-actions" slot="additional-actions"></slot>
        </kv-wizard-footer>
      </div>
    </div>
  </template>
</kv-wizard>
`;

exports[`Wizard (unit tests) when current step have success state should match the snapshot 1`] = `
<kv-wizard current-step="1" current-step-state="success" show-header="" show-step-bar="">
  <template shadowrootmode="open">
    <div class="wizard">
      <div class="wizard-header">
        <kv-wizard-header description="Configuration" label="Step 2">
          <slot name="additional-header-actions" slot="additional-header-actions"></slot>
        </kv-wizard-header>
      </div>
      <div class="has-header wizard-content">
        <slot name="step-content"></slot>
      </div>
      <div class="wizard-footer">
        <kv-wizard-footer cancelenabled="" completeenabled="" currentstep="1" exportparts="footer-actions-container" nextenabled="" prevenabled="" progresspercentage="50" shownextbtn="" showprevbtn="" showstepbar="">
          <slot name="additional-actions" slot="additional-actions"></slot>
        </kv-wizard-footer>
      </div>
    </div>
  </template>
</kv-wizard>
`;

exports[`Wizard (unit tests) when current step is the first one should match the snapshot 1`] = `
<kv-wizard current-step="0" current-step-state="success" show-header="" show-step-bar="">
  <template shadowrootmode="open">
    <div class="wizard">
      <div class="wizard-header">
        <kv-wizard-header description="Info" label="Step 1">
          <slot name="additional-header-actions" slot="additional-header-actions"></slot>
        </kv-wizard-header>
      </div>
      <div class="has-header wizard-content">
        <slot name="step-content"></slot>
      </div>
      <div class="wizard-footer">
        <kv-wizard-footer cancelenabled="" completeenabled="" currentstep="0" exportparts="footer-actions-container" nextenabled="" progresspercentage="0" showcancelbtn="" shownextbtn="" showstepbar="">
          <slot name="additional-actions" slot="additional-actions"></slot>
        </kv-wizard-footer>
      </div>
    </div>
  </template>
</kv-wizard>
`;

exports[`Wizard (unit tests) when current step is the last one should match the snapshot 1`] = `
<kv-wizard complete-btn-label="Deploy" current-step="2" current-step-state="success" show-header="" show-step-bar="">
  <template shadowrootmode="open">
    <div class="wizard">
      <div class="wizard-header">
        <kv-wizard-header description="Confirmation" label="Step 3">
          <slot name="additional-header-actions" slot="additional-header-actions"></slot>
        </kv-wizard-header>
      </div>
      <div class="has-header wizard-content">
        <slot name="step-content"></slot>
      </div>
      <div class="wizard-footer">
        <kv-wizard-footer cancelenabled="" completebtnlabel="Deploy" completeenabled="" currentstep="2" exportparts="footer-actions-container" nextenabled="" prevenabled="" progresspercentage="100" showcompletebtn="" showprevbtn="" showstepbar="">
          <slot name="additional-actions" slot="additional-actions"></slot>
        </kv-wizard-footer>
      </div>
    </div>
  </template>
</kv-wizard>
`;

exports[`Wizard (unit tests) when showHeader is false should match the snapshot 1`] = `
<kv-wizard current-step="1" show-step-bar="">
  <template shadowrootmode="open">
    <div class="wizard">
      <div class="wizard-content">
        <slot name="step-content"></slot>
      </div>
      <div class="wizard-footer">
        <kv-wizard-footer cancelenabled="" currentstep="1" exportparts="footer-actions-container" prevenabled="" progresspercentage="50" shownextbtn="" showprevbtn="" showstepbar="">
          <slot name="additional-actions" slot="additional-actions"></slot>
        </kv-wizard-footer>
      </div>
    </div>
  </template>
</kv-wizard>
`;

exports[`Wizard (unit tests) when showStepBar is false should match the snapshot 1`] = `
<kv-wizard current-step="1" show-header="">
  <template shadowrootmode="open">
    <div class="wizard">
      <div class="wizard-header">
        <kv-wizard-header description="Configuration" label="Step 2">
          <slot name="additional-header-actions" slot="additional-header-actions"></slot>
        </kv-wizard-header>
      </div>
      <div class="has-header wizard-content">
        <slot name="step-content"></slot>
      </div>
      <div class="wizard-footer">
        <kv-wizard-footer cancelenabled="" currentstep="1" exportparts="footer-actions-container" prevenabled="" progresspercentage="50" shownextbtn="" showprevbtn="">
          <slot name="additional-actions" slot="additional-actions"></slot>
        </kv-wizard-footer>
      </div>
    </div>
  </template>
</kv-wizard>
`;
